src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/ProjectTelemetry/ProjectLoadTelemetryReporter.cs

```diff
@@ -0,0 +1,110 @@
+                SdkStyleProject: isSdkStyleProject);
+
+            await ReportEventAsync(projectEvent, cancellationToken);
+        }
+        catch (Exception ex)
+        {
+            // Don't fail project loading because we failed to report telemetry.  Just log a warning and move on.
+            logger.LogWarning($"Failed to get project telemetry data: {ex.ToString()}");
+        }
+    }
+
+    private static async Task ReportEventAsync(ProjectLoadTelemetryEvent telemetryEvent, CancellationToken cancellationToken)
+    {
+        var instance = LanguageServerHost.Instance;
+        Contract.ThrowIfNull(instance, nameof(instance));
+        var clientLanguageServerManager = instance.GetRequiredLspService<IClientLanguageServerManager>();
+        await clientLanguageServerManager.SendNotificationAsync("workspace/projectConfiguration", telemetryEvent, cancellationToken);
+    }
+
+    private static ImmutableDictionary<HashedString, int> GetUniqueHashedFileExtensionsAndCounts(ProjectFileInfo projectFileInfo)
+    {
+        // Similar to O#, we report the content files + any non-generated source files.
+        var contentFiles = projectFileInfo.ProjectTelemetryMetadata.ContentFilePaths;
+        var sourceFiles = projectFileInfo.Documents
+            .Concat(projectFileInfo.AdditionalDocuments)
+            .Concat(projectFileInfo.AnalyzerConfigDocuments)
+            .Where(d => !d.IsGenerated)
+            .SelectAsArray(d => d.FilePath);
+        var allFiles = contentFiles.Concat(sourceFiles);
+        var filesCounts = allFiles.GroupBy(file => Path.GetExtension(file)).ToImmutableDictionary(kvp => VsTfmAndFileExtHashingAlgorithm.HashInput(kvp.Key), kvp => kvp.Count());
+        return filesCounts;
+    }
+
+    private static ImmutableArray<HashedString> GetHashedReferences(ImmutableArray<string> metadataReferences)
+    {
+        return metadataReferences.SelectAsArray(VsReferenceHashingAlgorithm.HashInput);
+    }
+
+    /// <summary>
+    /// This reads the solution file project id or hashes the contents+path
+    /// Matches O# implementation - https://github.com/OmniSharp/omnisharp-roslyn/blob/master/src/OmniSharp.MSBuild/ProjectLoadListener.cs#L88
+    /// </summary>
+    private static HashedString GetProjectId(ProjectToLoad projectToLoad)
+    {
+        if (projectToLoad.ProjectGuid is not null)
+        {
+            // The projectId is formatted as {GUID}.
+            // In order to match with O#, we need just the guid.
+            var projectGuid = projectToLoad.ProjectGuid.Replace("{", string.Empty).Replace("}", string.Empty);
+
+            // No need to actually hash the project guid.
+            return new HashedString(projectGuid);
+        }
+
+        var content = File.ReadAllText(projectToLoad.Path);
+        // This should exactly match O# to ensure we get the same hashes.
+        return VsReferenceHashingAlgorithm.HashInput($"Filename: {Path.GetFileName(projectToLoad.Path)}\n{content}");
```

Comment:
I'm scratching my head why this hash includes content which makes it vastly less stable than the GUID it's trying to act as a substitute for.

Reply:
Not sure.  I could modify it but you wouldn't be able to match project-only cases between O# / Roslyn.  That probably doesn't matter a lot I'm guessing (since this wasn't very stable to begin with).